### PR TITLE
Add `Grace_json_renderer`

### PR DESCRIPTION
**Pull request by WardBrian**
Closes #82.

For testing, I am re-using the ansi renderer tests to a large degree. Obviously they could be simplified, since the edge cases are much more limited in this renderer.

Example:
```bash
$ dune exec -- ./examples/main.exe --json
{
  "error_code": "E001",
  "severity": "error",
  "message": "`match` cases have incompatible types",
  "notes": [],
  "labels": [
    {
      "range": {
        "source": "fizz.ml",
        "start": { "line": 7, "column": 14 },
        "end": { "line": 8, "column": 1 }
      },
      "priority": "primary",
      "message": "expected `[> `Buzz | `Fizz | `Fizz_buzz]`, found `int`"
    },
    {
      "range": {
        "source": "fizz.ml",
        "start": { "line": 3, "column": 4 },
        "end": { "line": 8, "column": 1 }
      },
      "priority": "secondary",
      "message": "`match` cases have incompatible types"
    },
    {
      "range": {
        "source": "fizz.ml",
        "start": { "line": 4, "column": 14 },
        "end": { "line": 5, "column": 1 }
      },
      "priority": "secondary",
      "message": "this is found to be of type `[> `Fizz_buzz]`"
    },
    {
      "range": {
        "source": "fizz.ml",
        "start": { "line": 5, "column": 14 },
        "end": { "line": 6, "column": 1 }
      },
      "priority": "secondary",
      "message": "this is found to be of type `[> `Fizz]`"
    },
    {
      "range": {
        "source": "fizz.ml",
        "start": { "line": 6, "column": 14 },
        "end": { "line": 7, "column": 1 }
      },
      "priority": "secondary",
      "message": "this is found to be of type `[> `Buzz]`"
    }
  ]
}
```

### CODE REVIEW

File: lib/json_renderer/grace_json_renderer.mli [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+  :  ?code_to_string:('code -> string)
+  -> Format.formatter
+  -> 'code Grace.Diagnostic.t
+  -> unit
+
+val pp_diagnostics
+  :  ?code_to_string:('code -> string)
+  -> Format.formatter
+  -> 'code Grace.Diagnostic.t list
+  -> unit
```
Comment (johnyob) [inline lib/json_renderer/grace_json_renderer.mli]: Would it perhaps be more useful that the renderer returns the `Yojson.t` instead of printing it to a format buffer? In my mind, renderer doesn't have to imply printing here

**Review by johnyob — COMMENTED** [review]
Thanks for the contribution! 🙏 

Just a few comments (mostly style / nits), the logic all looks correct 🚀

**Review by johnyob — APPROVED** [review]
lgtm 🚀 

Seems like you need to make CI happy (likely your commit message format). After that we can merge :)

**Comment by jonsterling** [comment]
One thing I would like to raise is that AFAIK major language clients, such as that of Visual Studio Code, *only* properly support UTF-16 offsets. So integrating Grace with the language server protocol will in practice require switching to UTF-16 offsets, because VS Code is the de facto 'canonical' instance of LSP. 

Does it seem like there is a chance of making this change, or of at least providing this option? /cc @kentookura.
